1. Basics/task_2/hw2.php

```diff
@@ -0,0 +1,71 @@
+<?php
+/**
+ * @return string the input request example as a string
+ */
+function readHttpLikeInput() : string {
+    $f = fopen( 'php://stdin', 'r' );
+    $store = "";
+    $toread = 0;
+    while( $line = fgets( $f ) ) {
+        $store .= preg_replace("/\r/", "", $line);
+
+        // store a request body length if the body exsist
+        // $m - array of matches
+        // $m[1] <- (\d+); $m[1]*1 - (str->int) - the length of the body
+        if (preg_match('/Content-Length: (\d+)/',$line,$m)) {
+            $toread=$m[1]*1;
+        }
+
+        if ($line == "\r\n") 
+              break;
+    }
+
+    // if the request has a body, read and store it
+    if ($toread > 0) 
+        $store .= fread($f, $toread);
+    
+    return $store;
+}
+
+$contents = readHttpLikeInput();
+
+/**
+ * @param string $string : input request sample as a string
+ *
+ * @return array of the needed request parts
+ */
+function parseTcpStringAsHttpRequest($string) : array {
+    $strAsArr = explode("\n", $string);
+    $methodAndURI = explode(" ", array_shift($strAsArr));
+    $headers = [];
+    $body = NULL;
+
+    // if the request has headers store each header data as
+    // $headers[headerName] = headerValue
+    if($strAsArr) {
+        $i = 0;
+        while($strAsArr[$i]) {
+            $ind = strpos($strAsArr[$i], ":");
+            $headers[trim(substr($strAsArr[$i], 0, $ind))] = trim(substr($strAsArr[$i], $ind+1));
```

Comment:
if there is no `:` in the line (in body for example) - you will have incorrect behavior and errors.
if `strpos` doesn't find `:` - it returns `false`. The key will be empty string, but the value will be not "string", but "tring", as `false + 1 = 1`

Reply:
you use strpos too much. in this case there is an easier way to achieve what you want

Reply:
Do you mean split by ":" with ```explode()```? Or using regex` function?
All options are good for nice headers.
For bad header: "HelloIAmHeader: IHaveSomeMore_:_thanYouExpect ":
* strpos() will show junk,
* regex like ```'/^(.*):(.*)$/'``` may show junk,
*  expload() without some other checking will not show junk but smth like good one if 
``` php
$arr = explode(":", $badHeader);
$headers += [$arr[0] => $arr[1]];
```
It's better to see the junk after bad requests processig.

I guessed which option you mean?))